test/hotspot/jtreg/compiler/lib/compile_framework/Utils.java

```diff
@@ -0,0 +1,171 @@
+     * with multi-threading. Used to create the sources and classes directories. Since they
+     * are unique even across threads, the Compile Framework is multi-threading safe, i.e.
+     * it does not have collisions if two instances generate classes with the same name.
+     */
+    public static Path makeUniqueDir(String prefix) {
+        try {
+            return Files.createTempDirectory(Paths.get("."), prefix);
+        } catch (Exception e) {
+            throw new InternalCompileFrameworkException("Could not set up temporary directory", e);
+        }
+    }
+
+    /**
+     * Get all paths in the classpath.
+     */
+    public static String[] getClassPaths() {
+        String separator = File.pathSeparator;
+        return System.getProperty("java.class.path").split(separator);
+    }
+
+    /**
+     * Return the classpath, appended with the {@code classesDir}.
+     */
+    public static String getEscapedClassPathAndClassesDir(Path classesDir) {
+        String cp = System.getProperty("java.class.path") +
+                    File.pathSeparator +
+                    classesDir.toAbsolutePath();
+        // Escape the backslash for Windows paths. We are using the path in the
+        // command-line and Java code, so we always want it to be escaped.
+        return cp.replace("\\", "\\\\");
+    }
+
+    /**
+     * Get the path of asmtools, which is shipped with JTREG.
+     */
+    public static String getAsmToolsPath() {
+        for (String path : getClassPaths()) {
+            if (path.endsWith("jtreg.jar")) {
+                File jtreg = new File(path);
+                File dir = jtreg.getAbsoluteFile().getParentFile();
+                File asmtools = new File(dir, "asmtools.jar");
+                if (!asmtools.exists()) {
+                    throw new InternalCompileFrameworkException("Found jtreg.jar in classpath, but could not find asmtools.jar");
+                }
+                return asmtools.getAbsolutePath();
+            }
+        }
+        throw new InternalCompileFrameworkException("Could not find asmtools because could not find jtreg.jar in classpath");
+    }
+
+    private static void writeCodeToFile(String code, Path path) {
+        printlnVerbose("File: " + path);
+
+        // Ensure directory of the file exists.
+        Path dir = path.getParent();
+        try {
+            Files.createDirectories(dir);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not create directory: " + dir, e);
+        }
+
+        // Write to file.
+        try (BufferedWriter writer = Files.newBufferedWriter(path)) {
+            writer.write(code);
+        } catch (Exception e) {
+            throw new CompileFrameworkException("Could not write file: " + path, e);
+        }
+    }
+
+    /**
+     * Write sources to file.
+     */
+    public static List<Path> writeSourcesToFile(List<SourceCode> sources, Path sourceDir) {
```

Comment:
Since you write multiple files, should we name this "writeSourcesToFile**s**"?
```suggestion
    /**
     * Write each source in {@code sources} to a file inside {@code sourceDir}.
     */
    public static List<Path> writeSourcesToFiles(List<SourceCode> sources, Path sourceDir) {
```